pyat/at/physics/radiation.py

```diff
@@ -102,8 +102,10 @@ def process(r66):
             minv = inv(r66)
             r44 = inv(minv[:4, :4])
         # betatron emittances (dpp=0)
-        emit2 = numpy.sqrt(numpy.array(
-            [det(r44[s, s], check_finite=False) for s in _submat[:2]]))
+        emit2sq = numpy.array(
+            [det(r44[s, s], check_finite=False) for s in _submat[:2]])
+        # Prevent from unrealistic negative values of the determinant
+        emit2 = numpy.sqrt(numpy.maximum(emit2sq, 0.0))
```

Comment:
This doesn't fix the warning, we need to do the same for `emit3`:
```
emit3sq = numpy.array([det(r66[s, s]) for s in _submat])
# Prevent from unrealistic negative values of the determinant
emit3 = numpy.sqrt(numpy.maximum(emit3sq, 0.0))
```
If I understand the code correctly, I don't think we need to do it for `emit2` at all because accidentally negative values only occur for uncoupled machines, and now `emit3[1]` is not nan it should be caught by the eilf on line 97, please correct me if I am wrong.

Reply:
- You are right, the problem occurs only for uncoupled machines only, otherwise the vertical emittance is significantly positive.
- I do not think that the problem may occurs on `emit3` which comes directly from `r66` (at least this is not what happens with @carmignani ). It may come from the double matrix inversion at lines 98 and 100.
It's difficult to say since I cannot reproduce the problem, but I would keep the test as it is before adding another one on emit3, which is still possible later, but only if it is necessary.

Reply:
In @carmignani's test output you can see two warnings are raised:

```
test/test_physics.py::test_ohmi_envelope[refpts0]
test/test_physics.py::test_ohmi_envelope[refpts1]
  /mntdirect/_machfs/carmigna/python/at3.4/pyat/at/physics/radiation.py:87: RuntimeWarning: invalid value encountered in sqrt
    emit3 = numpy.sqrt(numpy.array([det(r66[s, s]) for s in _submat]))

test/test_physics.py::test_ohmi_envelope[refpts0]
test/test_physics.py::test_ohmi_envelope[refpts1]
  /mntdirect/_machfs/carmigna/python/at3.4/pyat/at/physics/radiation.py:100: RuntimeWarning: invalid value encountered in sqrt
    [det(r44[s, s], check_finite=False) for s in _submat[:2]]))
```

The first is for `emit3` the second is for `emit2`. I also had two warnings in my environment, after applying your fix I only had one, the `emit3` warning. The point I was trying to make above was that I thought that the only reason `emit2` was warning was because of the failure of `emit3`. I developed this theory after applying the fix to the `emit3` calculation but not the `emit2` calculation and both warnings disappeared for me. Perhaps I was a little overzealous in saying the check for `emit2` could be removed, but I definitely think that we need to add a check for `emit3`.

Reply:
However, I do not know which is the cause of the test failures as I am unable to run the tests in my environment that raises warnings. So patching just `emit2` my fix the tests but it certainly doesn't remove both warnings.

Reply:
You are right, I missed the warnings. But the 2 errors come from `emit2`.
`emit2` and `emit3` are two independent derivations from `r66`. Fixing one has no effect on the other. So I will add the fix on `emit3` also.

A word on the meaning of all this:
`r66` is a 6x6 symmetrical matrix which is the mathematical representation of a 6D ellipsoid with axes x, x', z, z', l, delta. To help in visualising this thing, we project the ellipsoid on 3 planes: x-x', z-z', l-delta. This gives 3 ellipses and their surfaces are the 3 values in emit3. For emit2, we first cut the ellipsoid along l=0 and delta=0. We are left with a 4D ellipsoid, which is projected on x-x' and z-z', giving 2 ellipses (contained in the ones of emit3, of course), and their surfaces in emit2.
In the case of an uncoupled machine, the ellipsoid is "flat", its volume is zero. When projected, we get on some planes (z-z' in our example) a segment of a line instead of an ellipse (an ellipse with surface 0). Because of rounding errors, this is even "thinner" than a line and the surface is imaginary. But for me, 0 or Nan mean the same: no vertical emittance. Clipping the emittance to zero is the best fix we can apply.

What worries me is the fact that I couldn't reproduce the bug on any machine, MacOS, Windows of Linux…

Reply:
Ahh ok, thank you for the explanation.
I agree, I found it very difficult to pin down even though I was able to replicate it in one environment.

Reply:
@lfarv I can now repeatably get warnings in any virtual environment by pip uninstalling numpy **and** scipy and then adding the locations of our Diamond versions to the python path. Will says that the only difference between them is that pip install uses wheels but the Diamond versions are source installed and they should be otherwise identical. Another thing worth noting is that `hmba.mat`, and so the tests, work fine for me but Diamond's `diad.mat` lattice raised warnings and had `nan` for `emitXY[1]`. Everything about this makes it look like it is going to be very hard to pin down a single root cause for the numerical errors. I think it is best to just settle with having fixed the problem on our end, that said I would be happy to test any theories you might have, now I can more easily replicate the conditions.